Clean/Circomlib/BinSub.lean

```diff
@@ -0,0 +1,147 @@
+import Clean.Circuit
+import Clean.Utils.Bits
+import Clean.Gadgets.Bits
+import Clean.Gadgets.Boolean
+
+
+
+namespace Circomlib
+open Utils.Bits
+variable {p : ℕ} [Fact p.Prime] [Fact (p > 2)]
+
+-- Instance to show that 2 > 0
+instance : Fact (0 < 2) := ⟨by norm_num⟩
+
+-- Define a 2D vector type for BinSub inputs
+-- Represents 2 operands, each with n bits
+-- This is a vector of 2 elements, where each element is a vector of n field elements
+@[reducible]
+def BinSubInput (n : ℕ) := ProvableVector (fields n) 2
+
+-- Define output type for BinSub
+-- Represents n output bits
+@[reducible]
+def BinSubOutput (n : ℕ) := fields n
+
+/-
+Original source code:
+https://github.com/iden3/circomlib/blob/master/circuits/binsub.circom
+
+The BinSub template takes two binary numbers as input and outputs their difference in binary form.
+The circuit computes: (in[0] + 2^n) - in[1] = out + aux*2^n
+where aux is the borrow bit.
+Note that the input bits must be guaranteed to be binary (0 or 1) by the caller.
+The circuit ensures that:
+1. All output bits are binary (0 or 1)
+2. The aux bit is binary (0 or 1)
+3. The equation (in[0] + 2^n) - in[1] = out + aux*2^n holds
+-/
+
+namespace BinSub
+/-
+template BinSub(n) {
+    signal input in[2][n];
+    signal output out[n];
+
+    signal aux;
+
+    var lin = 2**n;
+    var lout = 0;
+
+    var i;
+
+    for (i=0; i<n; i++) {
+        lin = lin + in[0][i]*(2**i);
+        lin = lin - in[1][i]*(2**i);
+    }
+
+    for (i=0; i<n; i++) {
+        out[i] <-- (lin >> i) & 1;
+
+        // Ensure out is binary
+        out[i] * (out[i] - 1) === 0;
+
+        lout = lout + out[i]*(2**i);
+    }
+
+    aux <-- (lin >> n) & 1;
+    aux*(aux-1) === 0;
+    lout = lout + aux*(2**n);
+
+    // Ensure the sum
+    lin === lout;
+}
+-/
+-- n: number of bits per operand
+def main (n : ℕ) [NeZero n] (inp : BinSubInput n (Expression (F p))) := do
+  -- Calculate input linear sum: lin = 2^n + in[0] - in[1]
+  let lin ← Circuit.foldlRange n ((2^n : F p) : Expression (F p)) fun lin i => do
+    let e2 : Expression (F p) := (2^i.val : F p)
+    return lin + inp[0][i] * e2 - inp[1][i] * e2
```

Comment:
this does no actual circuit operations (the only monad method used is `pure`), so it should be a plain definition `:=` instead of a monadic let. for example:
```suggestion
  let lin := Fin.foldl n (fun lin i =>
      let e2 : Expression (F p) := (2^i.val : F p)
      lin + inp[0][i] * e2 - inp[1][i] * e2)
    (2^n : F p)
```